We will be using SVC (Support Vector Classifier) from the Scikit-learn library to implement this model and evaluating the predictions using a classification report and a confusion matrix. We will later optimize the model using grid search.

In [1]:
##import libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.model_selection import GridSearchCV

##convert non-numeric variables
df = pd.read_csv('online_shoppers_intention.csv')
df



Administrative  Administrative_Duration  Informational  \
0                   0                      0.0              0   
1                   0                      0.0              0   
2                   0                      0.0              0   
3                   0                      0.0              0   
4                   0                      0.0              0   
...               ...                      ...            ...   
12325               3                    145.0              0   
12326               0                      0.0              0   
12327               0                      0.0              0   
12328               4                     75.0              0   
12329               0                      0.0              0   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                         0.0               1                 0.000000   
1                         0.0               2                64.000000   
2                         0.0               1                 0.000000   
3                         0.0               2                 2.666667   
4                         0.0              10               627.500000   
...                       ...             ...                      ...   
12325                     0.0              53              1783.791667   
12326                     0.0               5               465.750000   
12327                     0.0               6               184.250000   
12328                     0.0              15               346.000000   
12329                     0.0               3                21.250000   

       BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.200000   0.200000    0.000000         0.0   Feb                 1   
1         0.000000   0.100000    0.000000         0.0   Feb                 2   
2         0.200000   0.200000    0.000000         0.0   Feb                 4   
3         0.050000   0.140000    0.000000         0.0   Feb                 3   
4         0.020000   0.050000    0.000000         0.0   Feb                 3   
...            ...        ...         ...         ...   ...               ...   
12325     0.007143   0.029031   12.241717         0.0   Dec                 4   
12326     0.000000   0.021333    0.000000         0.0   Nov                 3   
12327     0.083333   0.086667    0.000000         0.0   Nov                 3   
12328     0.000000   0.021053    0.000000         0.0   Nov                 2   
12329     0.000000   0.066667    0.000000         0.0   Nov                 3   

       Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
0            1       1            1  Returning_Visitor    False    False  
1            2       1            2  Returning_Visitor    False    False  
2            1       9            3  Returning_Visitor    False    False  
3            2       2            4  Returning_Visitor    False    False  
4            3       1            4  Returning_Visitor     True    False  
...        ...     ...          ...                ...      ...      ...  
12325        6       1            1  Returning_Visitor     True    False  
12326        2       1            8  Returning_Visitor     True    False  
12327        2       1           13  Returning_Visitor     True    False  
12328        2       3           11  Returning_Visitor    False    False  
12329        2       1            2        New_Visitor     True    False  

[12330 rows x 18 columns]

In [3]:

from sklearn.preprocessing import LabelEncoder

# Define a list of features you want to encode
features_to_encode = ['Month', 'Weekend', 'Revenue', 'VisitorType']

# Initialize the LabelEncoder
label_encoder = LabelEncoder()

# Encode each feature in the DataFrame
for feature in features_to_encode:
    df[feature] = label_encoder.fit_transform(df[feature])
df
# If you want to transform new data using the same encoder,
##Set x and y variables


Administrative  Administrative_Duration  Informational  \
0                   0                      0.0              0   
1                   0                      0.0              0   
2                   0                      0.0              0   
3                   0                      0.0              0   
4                   0                      0.0              0   
...               ...                      ...            ...   
12325               3                    145.0              0   
12326               0                      0.0              0   
12327               0                      0.0              0   
12328               4                     75.0              0   
12329               0                      0.0              0   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                         0.0               1                 0.000000   
1                         0.0               2                64.000000   
2                         0.0               1                 0.000000   
3                         0.0               2                 2.666667   
4                         0.0              10               627.500000   
...                       ...             ...                      ...   
12325                     0.0              53              1783.791667   
12326                     0.0               5               465.750000   
12327                     0.0               6               184.250000   
12328                     0.0              15               346.000000   
12329                     0.0               3                21.250000   

       BounceRates  ExitRates  PageValues  SpecialDay  Month  \
0         0.200000   0.200000    0.000000         0.0      2   
1         0.000000   0.100000    0.000000         0.0      2   
2         0.200000   0.200000    0.000000         0.0      2   
3         0.050000   0.140000    0.000000         0.0      2   
4         0.020000   0.050000    0.000000         0.0      2   
...            ...        ...         ...         ...    ...   
12325     0.007143   0.029031   12.241717         0.0      1   
12326     0.000000   0.021333    0.000000         0.0      7   
12327     0.083333   0.086667    0.000000         0.0      7   
12328     0.000000   0.021053    0.000000         0.0      7   
12329     0.000000   0.066667    0.000000         0.0      7   

       OperatingSystems  Browser  Region  TrafficType  VisitorType  Weekend  \
0                     1        1       1            1            2        0   
1                     2        2       1            2            2        0   
2                     4        1       9            3            2        0   
3                     3        2       2            4            2        0   
4                     3        3       1            4            2        1   
...                 ...      ...     ...          ...          ...      ...   
12325                 4        6       1            1            2        1   
12326                 3        2       1            8            2        1   
12327                 3        2       1           13            2        1   
12328                 2        2       3           11            2        0   
12329                 3        2       1            2            0        1   

       Revenue  
0            0  
1            0  
2            0  
3            0  
4            0  
...        ...  
12325        0  
12326        0  
12327        0  
12328        0  
12329        0  

[12330 rows x 18 columns]

class  cap-diameter  cap-shape  cap-surface  cap-color  \
0         1         15.26          6            2          6   
1         1         16.60          6            2          6   
2         1         14.07          6            2          6   
3         1         14.17          2            3          1   
4         1         14.64          6            3          6   
...     ...           ...        ...          ...        ...   
9995      1          0.62          6            2          1   
9996      1          0.78          6            2          1   
9997      1          0.67          6            2          5   
9998      1          0.94          6            2          1   
9999      1          0.80          6            2          5   

      does-bruise-or-bleed  gill-attachment  gill-color  stem-height  \
0                        0                2          10        16.95   
1                        0                2          10        17.99   
2                        0                2          10        17.80   
3                        0                2          10        15.77   
4                        0                2          10        16.53   
...                    ...              ...         ...          ...   
9995                     0                0           7         3.89   
9996                     0                0           5         4.14   
9997                     0                0           7         4.10   
9998                     0                0           7         4.53   
9999                     0                0           5         4.00   

      stem-width  stem-color  has-ring  ring-type  habitat  season  
0          17.09          11         1          2        0       3  
1          18.19          11         1          2        0       2  
2          17.74          11         1          2        0       3  
3          15.98          11         1          5        0       3  
4          17.20          11         1          5        0       3  
...          ...         ...       ...        ...      ...     ...  
9995        0.98           4         0          1        0       1  
9996        1.03           4         0          1        0       1  
9997        1.00           4         0          1        0       2  
9998        1.19           4         0          1        0       0  
9999        1.00           4         0          1        0       2  

[10000 rows x 15 columns]

In [4]:
x=df.drop('Revenue', axis=1)
y=df['Revenue']

##Split the data into training and test tests
x_train, x_test, y_train, y_test =  train_test_split(x,y,test_size=0.3, random_state=12)

##set algorithm
model=SVC()

##fit the algorith to the training data
model.fit(x_train, y_train)

#assign a new variable to the prediction of the x test data
model_predict=model.predict(x_test)

#confusion matrix
print(confusion_matrix(y_test, model_predict))

#classification report
print(classification_report(y_test, model_predict))




[[3114    0]
 [ 580    5]]
              precision    recall  f1-score   support

           0       0.84      1.00      0.91      3114
           1       1.00      0.01      0.02       585

    accuracy                           0.84      3699
   macro avg       0.92      0.50      0.47      3699
weighted avg       0.87      0.84      0.77      3699



The confusion matrix reports no occurances of false positives (0), and the classification report states that precision, recall, and the f1-score are all above 0.84

Dear mr. Anderson The above Support vector matrix, is being used to analyze the predictability and consistancy that a customer makes a purchase given several potential purchase causing features. The data is reasonably significant with but strongly skewed toward predicting class 0 (no revenue) cases, than it is for predicting class 1 (Revenue) cases. In predicting no revenue it is accurate with an f1 score of 0.91, but has an f1 score as low as 0.02 meaning there is more data that will be needed for training for predicting the e of an unknown mushroom. It can be observed above that we've used a subset of the origional data for ease of processing, splitting a 10000 data subset, into a test and training pool. That said 7000 data points were used to train the model, and 3000 datapoints were used to test the model. in testing the model returned with precision and f1 scores both at 98% or higher for both positive and negative results. It should be noted that we are using only the first 10000 datapoints of a greater dataset which consists of 60000 datapoints, which is useable but would require a greater degree of processing time, which may be applied as needed for a final operation but as a proof of concept the above proves to be a valuable source for predicting edibility of a mushroom based on its given features. 